src/utils/view.test.js

```diff
@@ -0,0 +1,407 @@
+
+      const t1 = { x: 5, y: 10 };
+      setViewTransform(view, t1, 0, false);
+
+      const t2 = { x: 5, y: 20 };
+      setViewTransform(view, t2, 0, true);
+
+      const v = getViewTransform(view);
+      expect(v.x).toEqual(t1.x + t2.x);
+      expect(v.y).toEqual(t1.y + t2.y);
+      expect(v.k).toEqual(1);
+    });
+
+    it('setViewTransformExact: v = t', () => {
+      const view = mockView(0, 0);
+
+      const t = { x: 5, y: 10, k: 2 };
+      setViewTransformExact(view, t);
+
+      const v = getViewTransform(view);
+      expect(v).toEqual(t);
+    });
+
+    it('setViewTransform respects extents', () => {
+      const view = mockView(0, 0);
+
+      const extents = {
+        translate: {
+          minX: -110,
+          minY: -120,
+          maxX: 120,
+          maxY: 130
+        },
+        scale: {
+          minK: 1,
+          maxK: 1.5
+        }
+      };
+
+      setViewExtents(view, extents);
+
+      // Attempt going beyond max extents and max scale
+      setViewTransform(
+        view,
+        { x: 500, y: 500, k: extents.scale.maxK * 2 },
+        0,
+        false
+      );
+
+      // Result should have constrained to max extents * max scale
+      const v = getViewTransform(view);
+      expect(v.x).toEqual(extents.translate.maxX * extents.scale.maxK);
+      expect(v.y).toEqual(extents.translate.maxY * extents.scale.maxK);
+      expect(v.k).toEqual(extents.scale.maxK);
+
+      // Attempt going beyond min extents and min scale
+      setViewTransform(
+        view,
+        { x: -500, y: -500, k: extents.scale.minK * 0.5 },
+        0,
+        false
+      );
+
+      // Result should have constrained to min extents * min scale
+      const v2 = getViewTransform(view);
+      expect(v2.x).toEqual(extents.translate.minX * extents.scale.minK);
+      expect(v2.y).toEqual(extents.translate.minY * extents.scale.minK);
+      expect(v2.k).toEqual(extents.scale.minK);
+    });
+  });
+
+  describe('getViewExtents, setViewExtents', () => {
+    it('getViewExtents returns initial infinite extents', () => {
+      const view = mockView(100, 100);
+
+      expect(getViewExtents(view)).toEqual({
+        translate: {
+          minX: -Infinity,
+          maxX: Infinity,
+          minY: -Infinity,
+          maxY: Infinity
+        },
+        scale: {
+          minK: 0,
+          maxK: Infinity
+        }
+      });
+    });
+
+    it('setViewExtents applies given extents and getViewExtents returns same', () => {
+      const view = mockView(100, 100);
+
+      const extents = {
+        translate: {
+          minX: -1,
+          maxX: 2,
+          minY: -3,
+          maxY: 4
+        },
+        scale: {
+          minK: 0.5,
+          maxK: 0.6
+        }
+      };
+
+      setViewExtents(view, extents);
+      expect(getViewExtents(view)).toEqual(extents);
+    });
+  });
+
+  describe('viewTransformToFit', () => {
+    it('returns expected transform when example must be scaled down to fit', () => {
```

Comment:
It's hard to relate to which transforms scenerio these tests are supposed to mimic ( for example, I am guessing this one would likely refer to the opening of the metadata panel, but it is not clear to readers of the code). I understand the tests are structured similar to unit tests here, but perhaps they should be structued in a fashion that is easy for the reader to interpret as actual user scenerios?

Reply:
Yeah these are written as unit tests rather than end-to-end style which is why they feel a bit abstract. I know what you mean though I'll look into adding more tests on the flowchart that cover this in a more concrete way.